{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "07d08828-39be-453d-9c15-7d0737ceb72f",
   "metadata": {},
   "source": [
    "## Module_2:\n",
    "\n",
    "## Team Members:\n",
    "*Yaseen Noori and Marielle Miranda*\n",
    "\n",
    "## Project Title:\n",
    "From Image to Insight: Interpolative Prediction of Fibrotic Severity Across Lung Depths\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8372d796-186a-4ae2-9b4f-0065979e555a",
   "metadata": {},
   "source": [
    "## Project Goal:\n",
    "This project seeks to *develop an image analysis pipline that will predict the extent of fibrosis in the lung at different biopsy depths from the top of the lung* \n",
    "\n",
    "Two Main Questions:   \n",
    "        - How does the extent of fibrosis vary throughout the depth of a fibrtic lung?  \n",
    "        - Using interpolation, can you predict the amount of fibrosis at a specific depth into the lung?"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "91408b75-185a-4d66-9ce7-7c7aabbde5e1",
   "metadata": {},
   "source": [
    "## Disease Background: \n",
    "*Fill in information and please note that this module is truncated and only has 5 bullets (instead of the 11 that you did for Module #1).*\n",
    "\n",
    "* Prevalence & incidence: *In the United States there are around 140,000 patients living with idiopathic lung fibrosis and around 30,000 to 50,000 new cases diagnosed annually. Furthermore, worldwide, it is reported that 17.7 per 100,000 people are affected with idiopathic lung fibrosis and there are around 5.8  new cases per 100,000 people a year.*  \n",
    "    - https://medlineplus.gov/genetics/condition/idiopathic-pulmonary-fibrosis/#:~:text=Frequency,appear%20to%20run%20in%20families. \n",
    "    - https://pmc.ncbi.nlm.nih.gov/articles/PMC12330001/#:~:text=Pooled%20global%20incidence%20(5.8%20per compared%20with%20Europe%20and%20Asia. \n",
    "\n",
    "* Risk factors (genetic, lifestyle): *GENETIC factors that can lead to idiopathic lung fibrosis is a MUC5B gene variant (affects mucus clearing in lungs), surfactant genes (affects lung stability, such as SFTPC and SFTA1/2), telomere-related genes (can cause cell aging and poor lung repair. i.e. TERT, TERC, RTEL1), and family history (having family members with idiopathic lung fibrosis can increase risk). LIFESTYLE choices that can increase of getting idiopathic lung fibrosis are smoking, exposure to dust/fume, acid reflux/GERD (tiny amounts of acid that enter the lungs may cause injuries), viral infections (may trigger fibrosis), older age, and being a male (men have a greater risk of getting idiopathic lung fibrosis compared to women)*  \n",
    "    - ChatGPT Prompt: â€œWhat are the risk factors (genetic and lifestyle) of idiopathic lung fibrosis?\n",
    "\n",
    "* Symptoms: *Common symptoms for idiopathic lung fibrosis are the following: fatigue, difficulty breathing, persistent dry cough, and crackling sounds in the lungs (also known as â€œrailsâ€).* \n",
    "    - https://youtu.be/s2owdwl5Vac?si=7ZRDmlUugWh6iE28 \n",
    "\n",
    "* Standard of care treatment(s): *Common standard of care treatments for idiopathic lung fibrosis are antifibrotic medications, such as Nintedainb and Pirfenidone, which helps slow the disease progression by decreasing the formation of scar tissue through targeting fibroblasts and fibrocytes and interfering with signaling pathways. Another one is lifestyle adjustments, such as exercising and smoking cessation. If the progression of the disease is advanced, lung transplant may be an option or a mechanical ventilation.* \n",
    "    - https://www.mayoclinic.org/diseases-conditions/pulmonary-fibrosis/diagnosis-treatment/drc-20353695#:~:text=Physical%20exercise%20to%20improve%20how,treatment%20option%20for%20your%20condition. \n",
    "\n",
    "* Biological mechanisms (anatomy, organ physiology, cell & molecular physiology): *ANATOMY: Mainly, idiopathic lung fibrosis causes damage to the interstitial tissue (the thin layer between the alveoli and capillaries.) Over time, this tissue will thicken in the lower and outer regions of the lung (subpleural and basal). Furthermore, as the disease progresses, the normal lung structure becomes replaced with scar tissue and cystic spaces which create a â€œhoneycombâ€ look and reduces the overall size and elasticity of the lungs. ORGAN PHYSIOLOGY: With idiopathic lung fibrosis, the scar tissue causes the lungs to become stiff and less elastic, which makes it harder for the lung to expand when breathing. Furthermore, gas exchange becomes inefficient due to oxygen having to cross thickened and scared walls. Overall, this results in low blood oxygen, shortness of breath, and fatigue. If the progression gets worse, this may lead to the lung capacity decreasing and causing respiratory failure. CELL AND MOLECULAR PHYSIOLOGY: At the cellular level, idiopathic lung fibrosis is a result of abnormal wound healing in the alveoli after repeated microscopic injury. The damaged alveolar epithelial cells then release signals, such as TGF-Î², PDGF, and IL-13, which activate fibroblasts to produce excess collagen and scar tissue. Furthermore, due to idiopathic lung fibrosis, fibroblasts will keep multiplying while epithelial cells fail to regenerate. This leads to continuous fibrosis. Additionally, genetic factors (i.e. shortened telomeres and mitochondrial dysfunction) can lead to increased oxidative stress and can cause lung cells to be more vulnerable, while abnormal blood vessel growth and immune activation can further worsen the scarring process.* \n",
    "    - ChatGPT Prompt: â€œWhat are the biological mechanisms, such as anatomy, organ physiology, cell & molecular physiology, of idiopathic lung fibrosis?â€"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b99aefb-cb03-4bd8-b972-437eb0e02dfe",
   "metadata": {},
   "source": [
    "## Data-Set: \n",
    "* Describe the data set(s) you will analyze: *The data is 78 black and white images that are slices of mouse lungs at various depths. The black in the image is considered the healthy lungs, while the white is the fibrotic lesion that was caused from injecting the mouseâ€™s trachea with bleomycin to create lung fibrosis.*\n",
    "\n",
    "* Cite the source(s) of the data: *Module 2_Lecture 1 - 2025.pptx* \n",
    "\n",
    "* Describe how the data was collected -- What techniques were used?: *The data was collected through injecting bleomycin, an antibiotic known for being used in chemotherapy that can also cause lung fibrosis, into a mouseâ€™s trachea. After three weeks, the lung of the mouse is harvested and fixed in paraformaldehyde and mounted in gel or wax. Then using a cryotome or a microtome, the lungs are sliced into sections. The sections are then placed onto a glass microscopic slide and is immunostain by adding a fluorescent-labeled antibody that binds to the proteins (desmin smooth muscle alpha actin, and CD-31). Finally, the slices are viewed under a microscope and a digital image is taken. Important things to note is that the desmin signals are converted to black and white (white = fibrotic lesion and black = healthy lung) and there are a total of 78 images collected at various depths.* \n",
    "\n",
    "* What units are the data measured in?: *The data is measured in microns for the depth of the image.*\n",
    "\n",
    "* What subjects the data was obtained from?: *The subject the data was obtained from was one mouse that was injected with bleomycin in their trachea whose lungs were harvested after 3 weeks.*\n",
    "\n",
    "* Was there bias in how the data was collected?: *No, there are no biases in how the data was collected.* \n",
    "\n",
    "* What were the limitations/assumptions in the data set: *Yes, there are limitations and assumptions in the data set. One major limitation is that the study uses mouse lungs to model fibrosis in humans. Mouse lungs have a horizontal orientation, whereas human lungs have a vertical orientation. This difference alters gravitational, mechanical, and perfusion gradients, which can affect how fibrosis develops and distributes throughout the lung. Therefore, the results from mice may not fully replicate the spatial patterns or progression of fibrosis seen in humans.*\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4cde622-5508-4a54-aba8-77b454138bff",
   "metadata": {},
   "source": [
    "## Data Analyis: \n",
    "*(Describe how you analyzed the data. This is where you should intersperse your Python code so that anyone reading this can run your code to perform the analysis that you did, generate your figures, etc.)*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8f8df92",
   "metadata": {},
   "source": [
    "The following code is used to create a .csv file that contains the filename, depths, and percent white pixels for our 6 speciifc images that we chose: ch010017, ch010018, ch010019, ch010021, ch010022, and ch010159"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8acaf86b",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Module 2: count black and white pixels in a .jpg and extrapolate points'''\n",
    "\n",
    "from termcolor import colored          # lets us print colored text in the terminal (just for nicer output)\n",
    "import cv2                             # reads images and does basic image processing (we'll use thresholding)\n",
    "import numpy as np                     # helps with fast math on image arrays (counting pixels)\n",
    "import matplotlib.pyplot as plt        # makes plots (we'll show scatter plots)\n",
    "from scipy.interpolate import interp1d # makes a function that can estimate values between known data points\n",
    "import pandas as pd                    # makes a small table and writes it to a CSV file\n",
    "\n",
    "# Load the images you want to analyze \n",
    "# We'll point to each image by its full path. Order matters because it must match the depths list below.\n",
    "\n",
    "filenames = [\n",
    "    r\"/Users/mariellemiranda/Downloads/Module 2 - Lungs/MASK_Sk658 Llobe ch010017.jpg\",\n",
    "    r\"/Users/mariellemiranda/Downloads/Module 2 - Lungs/MASK_Sk658 Llobe ch010018.jpg\",\n",
    "    r\"/Users/mariellemiranda/Downloads/Module 2 - Lungs/MASK_Sk658 Llobe ch010019.jpg\",\n",
    "    r\"/Users/mariellemiranda/Downloads/Module 2 - Lungs/MASK_Sk658 Llobe ch010021.jpg\",\n",
    "    r\"/Users/mariellemiranda/Downloads/Module 2 - Lungs/MASK_Sk658 Llobe ch010022 (1).jpg\",\n",
    "    r\"/Users/mariellemiranda/Downloads/Module 2 - Lungs/MASK_SK658 Slobe ch010159.jpg\",\n",
    "]\n",
    "\n",
    "# Depths for each image in microns. The 1st number goes with the 1st filename, etc.\n",
    "# We'll use these as the x-axis values and for interpolation.\n",
    "depths = [\n",
    "    45,\n",
    "    90,\n",
    "    60,\n",
    "    30,\n",
    "    80,\n",
    "    860\n",
    "]\n",
    "\n",
    "# Empty lists we will fill as we go.\n",
    "images = []          # will store each image after we load it in grayscale\n",
    "white_counts = []    # number of white pixels (value 255) for each image\n",
    "black_counts = []    # number of black pixels (value 0) for each image\n",
    "white_percents = []  # percent of the image that is white (we'll compute this)\n",
    "\n",
    "# Read each image file in grayscale (0 = grayscale). Add it to our images list.\n",
    "for filename in filenames:\n",
    "    img = cv2.imread(filename, 0)   # if path is wrong, this could be None\n",
    "    images.append(img)\n",
    "\n",
    "# Now, for each image:\n",
    "# 1) Convert to pure black-and-white using a threshold\n",
    "# 2) Count white pixels and black pixels\n",
    "for x in range(len(filenames)):\n",
    "    # Any pixel >= 127 becomes 255 (white). Anything below 127 becomes 0 (black).\n",
    "    # We ignore the first returned value (the used threshold) with the underscore.\n",
    "    _, binary = cv2.threshold(images[x], 127, 255, cv2.THRESH_BINARY)\n",
    "\n",
    "    # Count how many pixels are white and black by checking equality in the binary image.\n",
    "    white = np.sum(binary == 255)\n",
    "    black = np.sum(binary == 0)\n",
    "\n",
    "    # Save those counts for this image.\n",
    "    white_counts.append(white)\n",
    "    black_counts.append(black)\n",
    "\n",
    "# Print out the raw counts so you can see them image by image.\n",
    "print(colored(\"Counts of pixel by color in each image\", \"yellow\"))\n",
    "for x in range(len(filenames)):\n",
    "    print(colored(f\"White pixels in image {x}: {white_counts[x]}\", \"white\"))\n",
    "    print(colored(f\"Black pixels in image {x}: {black_counts[x]}\", \"black\"))\n",
    "    print()  # just a blank line for readability\n",
    "\n",
    "# Turn counts into percentages: % white = 100 * white / (white + black).\n",
    "# We do this for every image and store the results.\n",
    "for x in range(len(filenames)):\n",
    "    white_percent = (100 * (white_counts[x] / (black_counts[x] + white_counts[x])))\n",
    "    white_percents.append(white_percent)\n",
    "\n",
    "# Show each filename and its % white along with its depth.\n",
    "print(colored(\"Percent white px:\", \"yellow\"))\n",
    "for x in range(len(filenames)):\n",
    "    print(colored(f'{filenames[x]}:', \"red\"))\n",
    "    print(f'{white_percents[x]}% White | Depth: {depths[x]} microns')\n",
    "    print()\n",
    "\n",
    "'''Write your data to a .csv file'''\n",
    "\n",
    "# Make a small table (DataFrame) that lines up each filename with its depth and percent white.\n",
    "df = pd.DataFrame({\n",
    "    'Filenames': filenames,\n",
    "    'Depths': depths,\n",
    "    'White percents': white_percents\n",
    "})\n",
    "\n",
    "# Save that table as Percent_White_Pixels.csv in whatever your current working folder is.\n",
    "# (If you don't see it, check the folder where you're running Python from.)\n",
    "df.to_csv('Percent_White_Pixels.csv', index=False)\n",
    "print(\"CSV file 'Percent_White_Pixels.csv' has been created.\")\n",
    "\n",
    "'''the .csv writing subroutine ends here'''\n",
    "\n",
    "# Ask the user for a depth value. We'll estimate (interpolate) the % white at that depth.\n",
    "# The input comes in as text, so we convert to float. 'colored' just makes the prompt yellow.\n",
    "interpolate_depth = float(input(colored(\"Enter the depth at which you want to interpolate a point: \", \"yellow\")))\n",
    "\n",
    "# x = known depths, y = known % white values. We'll build an interpolation function from these.\n",
    "x = depths\n",
    "y = white_percents\n",
    "\n",
    "# Make a simple straight-line (linear) interpolation function.\n",
    "# You could switch to 'quadratic' or 'cubic' if you had more data and wanted a curve.\n",
    "i = interp1d(x, y, kind='linear')\n",
    "\n",
    "# Use that function to estimate the % white at the depth the user typed in.\n",
    "interpolate_point = i(interpolate_depth)\n",
    "print(colored(f'The interpolated point is at the x-coordinate {interpolate_depth} and y-coordinate {interpolate_point}.', \"green\"))\n",
    "\n",
    "# For plotting, we make copies of the data lists and add the new (depth, %white) point to the end.\n",
    "depths_i = depths[:]\n",
    "depths_i.append(interpolate_depth)\n",
    "white_percents_i = white_percents[:]\n",
    "white_percents_i.append(interpolate_point)\n",
    "\n",
    "# Make a figure with 2 rows:\n",
    "# Top plot = only the original measured data\n",
    "# Bottom plot = original data + the new interpolated point highlighted in red\n",
    "fig, axs = plt.subplots(2, 1)\n",
    "\n",
    "# Plot 1: depth vs % white using your measured images\n",
    "axs[0].scatter(depths, white_percents, marker='o', linestyle='-', color='blue')\n",
    "axs[0].set_title('Plot of depth of image vs percentage white pixels')\n",
    "axs[0].set_xlabel('depth of image')\n",
    "axs[0].set_ylabel('white pixels as a percentage of total pixels')\n",
    "axs[0].grid(True)\n",
    "\n",
    "# Plot 2: same points, but also show the interpolated point at the end in red\n",
    "axs[1].scatter(depths_i, white_percents_i, marker='o', linestyle='-', color='blue')\n",
    "axs[1].set_title('Plot of depth of image vs percentage white pixels w/ interpolated point (red)')\n",
    "axs[1].set_xlabel('depth of image')\n",
    "axs[1].set_ylabel('white pixels as a percentage of total pixels')\n",
    "axs[1].grid(True)\n",
    "axs[1].scatter(\n",
    "    depths_i[len(depths_i)-1],                 # x-value of the last point (the new depth)\n",
    "    white_percents_i[len(white_percents_i)-1], # y-value of the last point (the new % white)\n",
    "    color='red', s=100, label='Highlighted point'\n",
    ")\n",
    "\n",
    "# Make spacing look nicer and show the plots.\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c67b6a4-ec94-4d28-b2a7-f6b860495118",
   "metadata": {},
   "source": [
    "## ðŸ§ª Verify and Validate Your Analysis\n",
    "\n",
    "**Verification (checking that the analysis ran correctly):**  \n",
    "We verified our analysis by confirming that each image was successfully loaded, converted to grayscale, and processed using OpenCVâ€™s `cv2.threshold()` function. For each image, we checked that both white and black pixel counts were non-zero and reflected visible differences in fibrosis severity. The white-pixel percentages printed in the terminal matched the values written to the generated `Percent_White_Pixels.csv` file, confirming that the output data was exported correctly.  \n",
    "\n",
    "We also tested the interpolation by entering known and intermediate depth values (for example, 200 Âµm and 500 Âµm) and verified that the predicted percent white pixels fell between the minimum and maximum measured values. The plots displayed clear data trends, and the interpolated point appeared in red at the correct position on the graphâ€”showing that the interpolation and plotting functions worked as intended.\n",
    "\n",
    "**Validation (checking that the results make biological sense):**  \n",
    "To validate our results, we compared our analysis trends with known biological findings from studies on bleomycin-induced pulmonary fibrosis in mouse models. Published research shows that fibrotic lesions increase in the lower and outer (subpleural and basal) regions of the lung, while upper regions remain relatively less affected. In our dataset, images taken at deeper depths of the lung showed higher percentages of white pixelsâ€”representing more fibrosisâ€”while shallower slices showed lower percentages. This pattern aligns with the expected physiological distribution of fibrosis, supporting that our analysis and interpolation pipeline produced results consistent with established biological evidence.\n",
    "\n",
    "**Sources:**  \n",
    "Moeller A, Ask K, Warburton D, Gauldie J, Kolb M. The bleomycin animal model: a useful tool to investigate treatment options for idiopathic pulmonary fibrosis? Int J Biochem Cell Biol. 2008;40(3):362-82. doi: 10.1016/j.biocel.2007.08.011. Epub 2007 Aug 30. PMID: 17936056; PMCID: PMC2323681.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2736cf95-2b93-444f-90c8-d40a54fc1df1",
   "metadata": {},
   "source": [
    "## ðŸ§­ Conclusions and Ethical Implications\n",
    "\n",
    "**Conclusions:**  \n",
    "Our image analysis and interpolation pipeline successfully quantified the extent of fibrosis across different lung depths. The results showed a clear pattern: deeper lung sections contained a higher percentage of white pixels, indicating more fibrosis, while shallower sections displayed less scarring. This supports our overarching question by demonstrating that fibrosis severity increases with tissue depth. The interpolation model also proved useful for estimating fibrosis at unmeasured depths, allowing for predictions without analyzing every physical tissue slice. Overall, this analysis shows that computational imaging can provide meaningful biological insight into disease progression.\n",
    "\n",
    "**Ethical Implications:**  \n",
    "Although our data came from animal experiments, it is important to acknowledge the ethical responsibilities involved. The induction of fibrosis using bleomycin must follow approved animal welfare and ethical research protocols (such as IACUC standards). On a positive note, developing digital image analysis tools like this can help reduce the need for additional animal testing in future studies."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f023b735-6efb-43ed-a03d-eb4a9cdb734e",
   "metadata": {},
   "source": [
    "## ðŸ”¬ Limitations and Future Work\n",
    "\n",
    "**Limitations:**  \n",
    "- Only six image slices were analyzed out of the full dataset of 78, which limits the accuracy and smoothness of the interpolation curve.  \n",
    "- The interpolation used a linear model, which assumes a straight-line relationship between fibrosis percentage and lung depth. In reality, fibrosis progression is likely nonlinear and may vary by region.  \n",
    "- The threshold value (127) used to separate black and white pixels was fixed and may not perfectly account for variations in lighting, staining, or image brightness, which can slightly affect pixel classification.  \n",
    "- The analysis only considered one channel (black and white) and did not include other staining information that could help differentiate tissue types or identify specific molecular markers.\n",
    "\n",
    "**Future Work:**  \n",
    "- Incorporate all 78 available images to build a more complete and detailed fibrosis-depth curve.  \n",
    "- Experiment with nonlinear interpolation methods (e.g., quadratic or cubic) to better capture complex changes in fibrosis severity.  \n",
    "- Use adaptive thresholding or machine learningâ€“based segmentation to improve accuracy under varying image conditions.  \n",
    "- Extend the analysis to include additional color or fluorescent channels to study relationships between fibrosis and other biological markers.  \n",
    "- Compare the interpolated fibrosis predictions with histopathological scores or published quantitative datasets to further validate the modelâ€™s accuracy.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9295960-2404-43dc-b46f-9a6f823f1657",
   "metadata": {},
   "source": [
    "## NOTES FROM YOUR TEAM: \n",
    "* *October 2 - We were given the data set and ask to create a .csv file for the percent of white pixels* \n",
    "* *October 7 - We found out how to convert a .ipynb to a .pdf and turned in our first notebook cheek*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd725b7f-9741-46b6-a213-9518da9201c3",
   "metadata": {},
   "source": [
    "## QUESTIONS FOR YOUR TA: \n",
    "*These are questions we have for our TA.* \n",
    "\n",
    "- Is there a scientific paper, like the last module, to learn more about the data set (i.e. data collection and the units the data is measured in)?\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.12 (vsc-bme)",
   "language": "python",
   "name": "vsc-bme"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
