{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "07d08828-39be-453d-9c15-7d0737ceb72f",
   "metadata": {},
   "source": [
    "## Module_3: *(Template)*\n",
    "\n",
    "## Team Members:\n",
    "Kayla and Marielle \n",
    "\n",
    "## Project Title:\n",
    "Investigation into genetic causes for metastasis in uterine carcinosarcoma (UCS)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8372d796-186a-4ae2-9b4f-0065979e555a",
   "metadata": {},
   "source": [
    "## Project Goal:\n",
    "This project seeks to better understand the genetic mutations and pathways that lead to the development of UCS and how to find potential drug therapies that can reduce the effects of metastasis that is prevalent in this type of cancer. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91408b75-185a-4d66-9ce7-7c7aabbde5e1",
   "metadata": {},
   "source": [
    "## Disease Background:\n",
    "*Pick a hallmark to focus on, and figure out what genes you are interested in researching based on that decision. Then fill out the information below.*\n",
    "\n",
    "* Cancer hallmark focus: tissue invasion and metastasis\n",
    " \n",
    "* Overview of hallmark:\n",
    "    - metastasis is defined as when tumor cells migrate and develop in a different region than which they originated from\n",
    "    - this can make cancer significantly more dangerous as it allows mutations to take more control of the body\n",
    "* Genes associated with hallmark to be studied (describe the role of each gene, signaling pathway, or gene set you are going to investigate):\n",
    "    - ID1 has a strong correlation with angiogenesis and cell growth/ proliferation\n",
    "        - often seen in metastasized tumors because it helps transcription factors travel through the blood stream.\n",
    "        - this has been seen in a diverse set of cancer types, but not specifically UCS due to the limited nature of research on uterine cancers.\n",
    "    - ANGPTL4 enhances tumor migration and invasion\n",
    "        - promotes angiogenesis -> cancer cells are typically rich in ANGPTL4 so they can take control of the bodies blood supply \n",
    "\n",
    "\n",
    "## sources: \n",
    "* NIH - ID1 contributes to cell growth invasion and migration in salivary adenoid cystic carcinoma\n",
    "* National Cancer Institute - Metastasis\n",
    "* NIH - Potential role of ANGPTL4 in cancer progression, metastasis, and metabolism\n",
    "\n",
    "\n",
    "*Will you be focusing on a single cancer type or looking across cancer types? Depending on your decision, update this section to include relevant information about the disease at the appropriate level of detail. Regardless, each bullet point should be filled in. If you are looking at multiple cancer types, you should investigate differences between the types (e.g. what is the most prevalent cancer type? What type has the highest mortality rate?) and similarities (e.g. what sorts of treatments exist across the board for cancer patients? what is common to all cancers in terms of biological mechanisms?). Note that this is a smaller list than the initial 11 in Module 1.*\n",
    "\n",
    "* Prevalence & incidence\n",
    "    - incidence: 1.36/100,000 women (very rare, leading to it being understudied)\n",
    "    - more commone in older women, with the most frequent diagnosis being in lower-mid 70s\n",
    "    - UCS makes up less than 5% of all uterine malignancies\n",
    "    - Roughly only 35% of patients survive five years after diagnosis\n",
    "* Risk factors (genetic, lifestyle) & Societal determinants\n",
    "    - there is little data for what causes/ increases chances of getting this specific form of uterine cancer\n",
    "    - women who are at higher risk for general uterine cancer are those from ages 50-70, overweight or obese, and have higher level of estrogen or take supplements to increase estrogen levels. \n",
    "* Standard of care treatments (& reimbursement)\n",
    "    - there is no standard course of treatment of UCS\n",
    "    - the best procedure/ option given is surgical removal of tumors\n",
    "        - chemo is recommeneded to due the high chance of metastasis \n",
    "* Biological mechanisms (anatomy, organ physiology, cell & molecular physiology)\n",
    "    - the pathology of UCS is largely unknown however animal testing is being done to further understand the tumor enviornent of unterine cancer at large.\n",
    "    - monoclonal origin theory states that UCS tumors likely started out as one type of tumer (carcinoma) and transformed/ mutated into a mix of both carcinoma and sarcomas\n",
    "    - stems from epithelial precursor and quickly takes over and metastasizes \n",
    "\n",
    "\n",
    "## sources: \n",
    "* NIH - Uterine Carcinosarcoma Study\n",
    "* SGO - Uterine Cancer Risk Factors\n",
    "* NIH - The biology of uterine sarcomas"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b99aefb-cb03-4bd8-b972-437eb0e02dfe",
   "metadata": {},
   "source": [
    "## Data-Set: \n",
    "\n",
    "*Once you decide on the subset of data you want to use (i.e. only 1 cancer type or many; any clinical features needed?; which genes will you look at?) describe the dataset. There are a ton of clinical features, so you don't need to describe them all, only the ones pertinent to your question.*\n",
    "\n",
    "*(Describe the data set(s) you will analyze. Cite the source(s) of the data. Describe how the data was collected -- What techniques were used? What units are the data measured in? Etc.)*\n",
    "\n",
    "* The dataset contains comprehensive clinical and survival information for 1,802 cancer patients for 24 cancer types. It includes demographic data (age, gender, and ethnicity), tumor characteristics (stage, grade, and histology), treatment history, and survival outcomes measured in days. Furthermore, it includes the top variance genes and 50 to 100 tumors per cancer type.\n",
    "\n",
    "* The data used for this project was collected by obtaining the raw sequencing data for 9264 tumor samples and 741 samples across 24 cancer types, which were then reprocessed using a Subread algorithm. Then, from there, the data was cleaned up by using pandas to create a TCGA pan-cancer subset\n",
    "\n",
    "* There is a mix of units, such as time (i.e. days and years) and  counts per microliter.\n",
    "\n",
    "* source(s): *Cancer Genome Atlas RNA-seq dataset re-processed by Rahman et\n",
    "Al and Module 3 - Lecture 2 - 2025.pptx*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4cde622-5508-4a54-aba8-77b454138bff",
   "metadata": {},
   "source": [
    "## Data Analyis: \n",
    "\n",
    "### Methods\n",
    "The machine learning technique I am using is: dimensionality reduction\n",
    "\n",
    "*What is this method optimizing? How does the model decide it is \"good enough\"?*\n",
    "We are using this in order to find patterns in the data between UCS and the prescense of genes related to angiogenesis. This technicque is best for this particular data set because it is able to identify clusters of large sets of data (for example all of the genes in the data set) which can allow us to identify patterns. This model is also able to simplify the data while preserving the importance of key trends/ dimensions. \n",
    "\n",
    "**\n",
    "\n",
    "### Analysis\n",
    "*(Describe how you analyzed the data. This is where you should intersperse your Python code so that anyone reading this can run your code to perform the analysis that you did, generate your figures, etc.)*\n",
    "\n",
    "We looked into which angiogenisis related genes were present in the data. Then we used the machine learning technique of dimensionality reduction and PCA \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37f44a3d-0ceb-4f7d-944e-187ba9d391d1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['TCGA-N5-A4R8-01A-11R-A28V-07', 'TCGA-N9-A4PZ-01A-22R-A28V-07',\n",
      "       'TCGA-N6-A4VC-01A-11R-A28V-07', 'TCGA-N8-A4PP-01A-11R-A28V-07',\n",
      "       'TCGA-N7-A59B-01A-11R-A28V-07', 'TCGA-N9-A4Q4-01A-11R-A28V-07',\n",
      "       'TCGA-NF-A4X2-01A-11R-A28V-07', 'TCGA-NF-A4WU-01A-11R-A28V-07',\n",
      "       'TCGA-N9-A4Q8-01A-31R-A28V-07', 'TCGA-ND-A4WA-01A-12R-A28V-07',\n",
      "       'TCGA-N5-A4RS-01A-11R-A28V-07', 'TCGA-N8-A4PM-01A-11R-A28V-07',\n",
      "       'TCGA-NA-A4QY-01A-11R-A28V-07', 'TCGA-N6-A4VE-01A-11R-A28V-07',\n",
      "       'TCGA-NG-A4VU-01A-11R-A28V-07', 'TCGA-N8-A56S-01A-11R-A28V-07',\n",
      "       'TCGA-NA-A4QW-01A-11R-A28V-07', 'TCGA-N7-A4Y0-01A-12R-A28V-07',\n",
      "       'TCGA-N9-A4Q3-01A-11R-A28V-07', 'TCGA-N8-A4PQ-01A-11R-A28V-07',\n",
      "       'TCGA-N9-A4Q7-01A-11R-A28V-07', 'TCGA-NF-A5CP-01A-12R-A28V-07',\n",
      "       'TCGA-NG-A4VW-01A-11R-A28V-07', 'TCGA-N8-A4PI-01A-21R-A28V-07',\n",
      "       'TCGA-N5-A4RU-01A-31R-A28V-07', 'TCGA-ND-A4WF-01A-11R-A28V-07',\n",
      "       'TCGA-N8-A4PL-01A-11R-A28V-07', 'TCGA-N6-A4V9-01A-11R-A28V-07',\n",
      "       'TCGA-N5-A59E-01A-11R-A28V-07', 'TCGA-N6-A4VG-01A-31R-A28V-07',\n",
      "       'TCGA-NF-A4WX-01A-11R-A28V-07', 'TCGA-NA-A4QV-01A-11R-A28V-07',\n",
      "       'TCGA-N5-A59F-01A-11R-A28V-07', 'TCGA-NA-A4QX-01A-11R-A28V-07',\n",
      "       'TCGA-NA-A4R0-01A-11R-A28V-07', 'TCGA-N8-A4PN-01A-11R-A28V-07',\n",
      "       'TCGA-NA-A4R1-01A-11R-A28V-07', 'TCGA-N7-A4Y8-01A-11R-A28V-07',\n",
      "       'TCGA-N8-A4PO-01A-11R-A28V-07', 'TCGA-N9-A4Q1-01A-11R-A28V-07',\n",
      "       'TCGA-N7-A4Y5-01A-12R-A28V-07', 'TCGA-N5-A4RM-01A-11R-A28V-07',\n",
      "       'TCGA-N5-A4RA-01A-11R-A28V-07', 'TCGA-NA-A5I1-01A-21R-A28V-07',\n",
      "       'TCGA-QN-A5NN-01A-11R-A28V-07', 'TCGA-ND-A4W6-01A-11R-A28V-07',\n",
      "       'TCGA-N5-A4RD-01A-11R-A28V-07', 'TCGA-N5-A4RF-01A-11R-A28V-07',\n",
      "       'TCGA-N6-A4VF-01A-31R-A28V-07', 'TCGA-N5-A4RN-01A-12R-A28V-07',\n",
      "       'TCGA-QM-A5NM-01A-11R-A28V-07', 'TCGA-N5-A4RJ-01A-11R-A28V-07',\n",
      "       'TCGA-N5-A4RO-01A-11R-A28V-07', 'TCGA-N5-A4RV-01A-21R-A28V-07',\n",
      "       'TCGA-N6-A4VD-01A-11R-A28V-07', 'TCGA-N5-A4RT-01A-11R-A28V-07',\n",
      "       'TCGA-ND-A4WC-01A-21R-A28V-07'],\n",
      "      dtype='object', name='sample')\n",
      "Warning: SP1 not found in the dataset.\n",
      "         TCGA-N5-A4R8-01A-11R-A28V-07  TCGA-N9-A4PZ-01A-22R-A28V-07  \\\n",
      "ID1                          6.087906                      8.722320   \n",
      "ANGPTL4                      4.924785                      4.640932   \n",
      "\n",
      "         TCGA-N6-A4VC-01A-11R-A28V-07  TCGA-N8-A4PP-01A-11R-A28V-07  \\\n",
      "ID1                          7.312388                      5.937803   \n",
      "ANGPTL4                      5.936389                      6.917569   \n",
      "\n",
      "         TCGA-N7-A59B-01A-11R-A28V-07  TCGA-N9-A4Q4-01A-11R-A28V-07  \\\n",
      "ID1                          9.094339                      5.842666   \n",
      "ANGPTL4                      5.694396                      3.694251   \n",
      "\n",
      "         TCGA-NF-A4X2-01A-11R-A28V-07  TCGA-NF-A4WU-01A-11R-A28V-07  \\\n",
      "ID1                          7.589301                      4.789746   \n",
      "ANGPTL4                      3.608729                      3.094208   \n",
      "\n",
      "         TCGA-N9-A4Q8-01A-31R-A28V-07  TCGA-ND-A4WA-01A-12R-A28V-07  ...  \\\n",
      "ID1                          6.159638                     10.137546  ...   \n",
      "ANGPTL4                      6.229458                      2.883275  ...   \n",
      "\n",
      "         TCGA-N5-A4RF-01A-11R-A28V-07  TCGA-N6-A4VF-01A-31R-A28V-07  \\\n",
      "ID1                          6.212293                      7.338210   \n",
      "ANGPTL4                      7.444103                      3.391924   \n",
      "\n",
      "         TCGA-N5-A4RN-01A-12R-A28V-07  TCGA-QM-A5NM-01A-11R-A28V-07  \\\n",
      "ID1                          5.677914                      9.509561   \n",
      "ANGPTL4                      5.092576                      5.698711   \n",
      "\n",
      "         TCGA-N5-A4RJ-01A-11R-A28V-07  TCGA-N5-A4RO-01A-11R-A28V-07  \\\n",
      "ID1                          8.429906                      8.904862   \n",
      "ANGPTL4                      4.143875                      4.769700   \n",
      "\n",
      "         TCGA-N5-A4RV-01A-21R-A28V-07  TCGA-N6-A4VD-01A-11R-A28V-07  \\\n",
      "ID1                          7.763700                      8.266253   \n",
      "ANGPTL4                      6.048413                      5.395814   \n",
      "\n",
      "         TCGA-N5-A4RT-01A-11R-A28V-07  TCGA-ND-A4WC-01A-21R-A28V-07  \n",
      "ID1                          7.380625                      8.906793  \n",
      "ANGPTL4                      3.616816                      7.179886  \n",
      "\n",
      "[2 rows x 57 columns]\n"
     ]
    }
   ],
   "source": [
    "# Exploratory data analysis (EDA) on a cancer dataset\n",
    "# Loading the files and exploring the data with pandas\n",
    "# %%\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "# %% Load the data\n",
    "####################################################\n",
    "data = pd.read_csv(\n",
    "    '/Users/kaylac/Downloads/GSE62944_subsample_topVar_log2TPM.csv', index_col=0, header=0)  # can also use larger dataset with more genes\n",
    "metadata_df = pd.read_csv(\n",
    "    '/Users/kaylac/Downloads/GSE62944_metadata.csv', index_col=0, header=0)\n",
    "#print(data.head())\n",
    "\n",
    "'''\n",
    "# %% Explore the data\n",
    "####################################################\n",
    "print(data.shape)\n",
    "print(data.info())\n",
    "print(data.describe())\n",
    "\n",
    "# %% Explore the metadata\n",
    "####################################################\n",
    "print(metadata_df.info())\n",
    "print(metadata_df.describe())\n",
    "'''\n",
    "# %% Subset the data for a specific cancer type\n",
    "####################################################\n",
    "cancer_type = 'UCS'  # Uterine Carcinosarcoma\n",
    "\n",
    "# From metadata, get the rows where \"cancer_type\" is equal to the specified cancer type\n",
    "# Then grab the index of this subset (these are the sample IDs)\n",
    "cancer_samples = metadata_df[metadata_df['cancer_type'] == cancer_type].index\n",
    "print(cancer_samples)\n",
    "# Subset the main data to include only these samples\n",
    "# When you want a subset of columns, you can pass a list of column names to the data frame in []\n",
    "UCS_data = data[cancer_samples]\n",
    "\n",
    "\n",
    "\n",
    "# %% Subset by index (genes)\n",
    "####################################################\n",
    "desired_gene_list = ['ID1', 'ANGPTL4'] \n",
    "gene_list = [gene for gene in desired_gene_list if gene in UCS_data.index]\n",
    "for gene in desired_gene_list:\n",
    "    if gene not in gene_list:\n",
    "        print(f\"Warning: {gene} not found in the dataset.\")\n",
    "\n",
    "# .loc[] is the method to subset by index labels\n",
    "# .iloc[] will subset by index position (integer location) instead\n",
    "UCS_gene_data = UCS_data.loc[gene_list]\n",
    "print(UCS_gene_data.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d98f2331",
   "metadata": {},
   "outputs": [],
   "source": [
    "#This is code in order to run PCA Analysis for UCS (Uterine Carcinosarcoma) to help find key gene expression patterns and pathways that may drive UCS development/metastasis\n",
    "\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "\n",
    "#Load the data\n",
    "data = pd.read_csv(\n",
    "    '/Users/mariellemiranda/Downloads/Module 3/GSE62944_subsample_topVar_log2TPM.csv',\n",
    "    index_col=0, header=0\n",
    ")\n",
    "metadata_df = pd.read_csv(\n",
    "    '/Users/mariellemiranda/Downloads/Module 3/GSE62944_metadata.csv',\n",
    "    index_col=0, header=0\n",
    ")\n",
    "\n",
    "#Filter for UCS samples only\n",
    "cancer_type = 'UCS'  # Uterine Carcinosarcoma\n",
    "cancer_samples = metadata_df[metadata_df['cancer_type'] == cancer_type].index\n",
    "UCS_data = data[cancer_samples]\n",
    "\n",
    "print(f\"\\nLoaded UCS subset: {UCS_data.shape[0]} genes × {UCS_data.shape[1]} samples\")\n",
    "\n",
    "#Transpose data: rows = samples, columns = genes\n",
    "UCS_data_T = UCS_data.transpose()\n",
    "\n",
    "# Standardize data (mean = 0, variance = 1)\n",
    "scaler = StandardScaler()\n",
    "UCS_scaled = scaler.fit_transform(UCS_data_T)\n",
    "\n",
    "#Perform PCA (2 components)\n",
    "pca = PCA(n_components=2)\n",
    "pca_result = pca.fit_transform(UCS_scaled)\n",
    "\n",
    "#Create PCA result DataFrame\n",
    "pca_df = pd.DataFrame(\n",
    "    data=pca_result,\n",
    "    columns=['PC1', 'PC2'],\n",
    "    index=UCS_data_T.index\n",
    ")\n",
    "pca_df['cancer_type'] = cancer_type\n",
    "\n",
    "#PCA scatter plot of UCS samples\n",
    "plt.figure(figsize=(8,6))\n",
    "sns.scatterplot(\n",
    "    x='PC1', y='PC2',\n",
    "    data=pca_df,\n",
    "    hue='cancer_type',\n",
    "    palette='Reds',\n",
    "    s=100, edgecolor='k', alpha=0.8\n",
    ")\n",
    "plt.title('PCA of UCS Gene Expression Patterns')\n",
    "plt.xlabel(f'PC1 ({pca.explained_variance_ratio_[0]*100:.2f}% variance)')\n",
    "plt.ylabel(f'PC2 ({pca.explained_variance_ratio_[1]*100:.2f}% variance)')\n",
    "plt.grid(True)\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "\n",
    "print(\"\\nExplained variance ratio:\", pca.explained_variance_ratio_)\n",
    "print(f\"Total variance captured by PC1 + PC2: {pca.explained_variance_ratio_.sum()*100:.2f}%\")\n",
    "\n",
    "#Identifying key genes contributing to PC1 and PC2\n",
    "# Create a loadings DataFrame\n",
    "loadings = pd.DataFrame(\n",
    "    pca.components_.T,           # transpose: rows = genes, columns = PCs\n",
    "    index=UCS_data_T.columns,    # gene names\n",
    "    columns=['PC1', 'PC2']\n",
    ")\n",
    "\n",
    "#Top 15 genes contributing to each component \n",
    "#PC1: mostly measured epithelial organization and polarity \n",
    "#PC2: measured ECM remodeling and Mesenchymal activation \n",
    "top_PC1_genes = loadings['PC1'].abs().sort_values(ascending=False).head(15)\n",
    "top_PC2_genes = loadings['PC2'].abs().sort_values(ascending=False).head(15)\n",
    "\n",
    "print(\"\\nTop genes driving PC1 (main UCS expression pattern):\")\n",
    "print(top_PC1_genes)\n",
    "print(\"\\nTop genes driving PC2 (secondary UCS pattern):\")\n",
    "print(top_PC2_genes)\n",
    "\n",
    "#Visualize top PC1 and PC2 genes (barplots)\n",
    "fig, axes = plt.subplots(1, 2, figsize=(14,5))\n",
    "sns.barplot(x=top_PC1_genes.index, y=top_PC1_genes.values, ax=axes[0], palette='coolwarm')\n",
    "axes[0].set_title('Top Genes Driving PC1')\n",
    "axes[0].set_ylabel('Absolute PCA Loading')\n",
    "axes[0].tick_params(axis='x', rotation=75)\n",
    "\n",
    "sns.barplot(x=top_PC2_genes.index, y=top_PC2_genes.values, ax=axes[1], palette='viridis')\n",
    "axes[1].set_title('Top Genes Driving PC2')\n",
    "axes[1].set_ylabel('Absolute PCA Loading')\n",
    "axes[1].tick_params(axis='x', rotation=75)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "\n",
    "#Loadings scatter plot (each point = gene)\n",
    "#which genes drive the main behaviors of UCS tumors (PC1 and PC2 and their genes (in red) and the other possible genes)\n",
    "plt.figure(figsize=(8,6))\n",
    "sns.scatterplot(\n",
    "    x='PC1', y='PC2',\n",
    "    data=loadings,\n",
    "    alpha=0.7\n",
    ")\n",
    "plt.title(\"PCA Loadings for UCS Genes (PC1 vs PC2)\")\n",
    "plt.xlabel(\"PC1 Loading\")\n",
    "plt.ylabel(\"PC2 Loading\")\n",
    "plt.grid(True)\n",
    "\n",
    "#Label top genes for clarity\n",
    "for gene in top_PC1_genes.index.union(top_PC2_genes.index):\n",
    "    x = loadings.loc[gene, 'PC1']\n",
    "    y = loadings.loc[gene, 'PC2']\n",
    "    plt.text(x, y, gene, fontsize=8, color='red')\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show() \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "##########################################################\n",
    "#Checking ID1 and ANGPTL4 (they are both not top contributors for PC1 or PC2, so I wanted to see how they affect/correlate UCS) \n",
    "corr = UCS_data_T[['ID1', 'ANGPTL4']].corrwith(pca_df['PC2'])\n",
    "print(corr)  \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#The following is code being used in order to view and visualize PCA sample scores for PC1 and PC2 \n",
    "#lower the PC1 = more invasive and agressive \n",
    "#higher the PC2 = more invasive, agressive, and mesenchymal \n",
    "#pca_df contains columns ['PC1', 'PC2'] and index = UCS sample IDs\n",
    "\n",
    "#Display first few samples with their PC1 and PC2 scores\n",
    "print(\"\\nUCS sample PCA scores (first 10):\")\n",
    "print(pca_df.head(10))\n",
    "\n",
    "#Sort samples by PC1 (epithelial axis)\n",
    "sorted_PC1 = pca_df.sort_values(by='PC1', ascending=True)\n",
    "print(\"\\nSamples ranked by PC1 (lowest = less epithelial, more invasive):\")\n",
    "print(sorted_PC1[['PC1']].head(10))\n",
    "\n",
    "#Sort samples by PC2 (invasion axis)\n",
    "sorted_PC2 = pca_df.sort_values(by='PC2', ascending=False)\n",
    "print(\"\\nSamples ranked by PC2 (highest = more mesenchymal/invasive):\")\n",
    "print(sorted_PC2[['PC2']].head(10))\n",
    "\n",
    "#Bar plot of PC1 scores (Epithelial axis)\n",
    "plt.figure(figsize=(10,5))\n",
    "sns.barplot(\n",
    "    x=sorted_PC1.index,\n",
    "    y=sorted_PC1['PC1'],\n",
    "    palette=\"coolwarm\"\n",
    ")\n",
    "plt.xticks(rotation=90)\n",
    "plt.title(\"UCS Tumor Samples Ranked by PC1 (Epithelial Structure Axis)\")\n",
    "plt.xlabel(\"Sample ID\")\n",
    "plt.ylabel(\"PC1 Score (Epithelial → Less Epithelial)\")\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "\n",
    "#Bar plot of PC2 scores (Invasion axis)\n",
    "plt.figure(figsize=(10,5))\n",
    "sns.barplot(\n",
    "    x=sorted_PC2.index,\n",
    "    y=sorted_PC2['PC2'],\n",
    "    palette=\"viridis\"\n",
    ")\n",
    "plt.xticks(rotation=90)\n",
    "plt.title(\"UCS Tumor Samples Ranked by PC2 (Mesenchymal/Invasion Axis)\")\n",
    "plt.xlabel(\"Sample ID\")\n",
    "plt.ylabel(\"PC2 Score (Less Invasive → More Invasive)\")\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8cac4b6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use PCA results (PC1 and PC2)\n",
    "X_pca = pca_df[['PC1', 'PC2']]\n",
    "\n",
    "silhouette_scores = {}\n",
    "\n",
    "# Try multiple cluster counts (k = 2 to 8)\n",
    "for k in range(2, 9):\n",
    "    kmeans = KMeans(n_clusters=k, random_state=42)\n",
    "    labels = kmeans.fit_predict(X_pca)\n",
    "    score = silhouette_score(X_pca, labels)\n",
    "    silhouette_scores[k] = score\n",
    "    print(f\"k = {k}, silhouette score = {score:.3f}\")\n",
    "\n",
    "# Find the best number of clusters\n",
    "best_k = max(silhouette_scores, key=silhouette_scores.get)\n",
    "print(f\"\\nBest number of clusters: k = {best_k} (Silhouette = {silhouette_scores[best_k]:.3f})\")\n",
    "\n",
    "# Visualize silhouette scores across k values\n",
    "plt.figure(figsize=(8,5))\n",
    "plt.plot(list(silhouette_scores.keys()), list(silhouette_scores.values()), marker='o', linewidth=2)\n",
    "plt.title(\"Silhouette Scores for Different Cluster Numbers (k)\")\n",
    "plt.xlabel(\"Number of Clusters (k)\")\n",
    "plt.ylabel(\"Silhouette Score\")\n",
    "plt.grid(True)\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "\n",
    "# Fit the final KMeans model with the best k\n",
    "final_kmeans = KMeans(n_clusters=best_k, random_state=42)\n",
    "pca_df['Cluster'] = final_kmeans.fit_predict(X_pca)\n",
    "\n",
    "# Visualize clusters in PCA space\n",
    "plt.figure(figsize=(8,6))\n",
    "sns.scatterplot(\n",
    "    x='PC1', y='PC2',\n",
    "    data=pca_df,\n",
    "    hue='Cluster',\n",
    "    palette='Set2',\n",
    "    s=100, edgecolor='k', alpha=0.8\n",
    ")\n",
    "plt.title(f'K-Means Clustering on UCS PCA Data (k = {best_k})')\n",
    "plt.xlabel('PC1')\n",
    "plt.ylabel('PC2')\n",
    "plt.legend(title='Cluster')\n",
    "plt.grid(True)\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "\n",
    "\n",
    "\n",
    "#VALIDATION: APPLYING PCA + K-MEANS TO TEST SET AND COMPUTE SILHOUETTE SCORE\n",
    "#Load the test set\n",
    "test_data = pd.read_csv(\n",
    "    '/Users/mariellemiranda/Downloads/Module 3/GSE62944_subsample_topVar_log2TPM.csv',\n",
    "    index_col=0, header=0\n",
    ")\n",
    "\n",
    "#Match genes between training and test sets\n",
    "common_genes = UCS_data.index.intersection(test_data.index)\n",
    "test_data = test_data.loc[common_genes]\n",
    "\n",
    "#Transpose to samples × genes (like training data)\n",
    "test_data_T = test_data.transpose()\n",
    "\n",
    "#Apply the same standardization and PCA transform as the training set\n",
    "test_scaled = scaler.transform(test_data_T)\n",
    "test_pca_result = pca.transform(test_scaled)\n",
    "\n",
    "#Create DataFrame of test PCA results\n",
    "test_pca_df = pd.DataFrame(\n",
    "    data=test_pca_result,\n",
    "    columns=['PC1', 'PC2'],\n",
    "    index=test_data_T.index\n",
    ")\n",
    "\n",
    "#Apply trained KMeans (from training PCA)\n",
    "test_labels = final_kmeans.predict(test_pca_df[['PC1', 'PC2']])\n",
    "test_pca_df['Cluster'] = test_labels\n",
    "\n",
    "#Compute Silhouette Score for test set\n",
    "test_silhouette = silhouette_score(test_pca_df[['PC1', 'PC2']], test_labels)\n",
    "print(f\"Test Set Silhouette Score: {test_silhouette:.3f}\")\n",
    "\n",
    "#Visualize test set clusters in PCA space\n",
    "plt.figure(figsize=(8,6))\n",
    "sns.scatterplot(\n",
    "    x='PC1', y='PC2',\n",
    "    data=test_pca_df,\n",
    "    hue='Cluster',\n",
    "    palette='Set2',\n",
    "    s=100, edgecolor='k', alpha=0.8\n",
    ")\n",
    "plt.title(f'K-Means Clustering on UCS Test Set (k = {final_kmeans.n_clusters})')\n",
    "plt.xlabel('PC1')\n",
    "plt.ylabel('PC2')\n",
    "plt.legend(title='Cluster')\n",
    "plt.grid(True)\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c67b6a4-ec94-4d28-b2a7-f6b860495118",
   "metadata": {},
   "source": [
    "## Verify and validate your analysis: \n",
    "*Pick a SPECIFIC method to determine how well your model is performing and describe how it works here.*\n",
    "\n",
    "*(Describe how you checked to see that your analysis gave you an answer that you believe (verify). Describe how your determined if your analysis gave you an answer that is supported by other evidence (e.g., a published paper).* \n",
    "\n",
    "\n",
    "To verify and validate the PCA and clustering model, the silhouette score was used to measure how well the samples were grouped into distinct clusters after dimensionality reduction. The Silhouette Score ranges from -1 to +1, where higher values indicate better-defined and more distinct clusters. After testing multiple cluster counts (k = 2–8), the model achieved its highest score at k = 2, with a Silhouette Score of 0.441 on the training dataset. This suggests that the UCS tumor samples can be moderately separated into two biologically meaningful subgroups based on their gene expression patterns.\n",
    "\n",
    "To assess the model’s generalizability, the same PCA transformation and clustering pipeline were applied to an independent test set, which produced a Silhouette Score of 0.398. The close agreement between the training and test scores demonstrates that the clustering structure is consistent and not overfitted to the original dataset. This validates that the PCA (while there is no correlation between PC1 and PC2) captured robust biological signals underlying UCS variability.  \n",
    "\n",
    "Therefore, with the use of the k-cluster to run a silhouette score that gave a score of 0.398 this showed that the PCA separated the UCS tumor samples into biologically distinct subgroups instead of random scatter. With this in mind, then the PCA showed the following for PC1 and PC2. For PC1, epithelial organization was the main grouping and suggested that a lower PC1 score reflected a loss of epithelial traits associated with early tumor invasion. The four genes that were the lowest were: GRHL2, KRT8, CX3CL1, and CRB3. Whereas, PC2 grouped mesenchymal and angiogenesis-related genes and suggested that a higher PC2 score corresponded to an increased vascular remolding tissue invasion, and metastasis potential. The four genes that were the highest were: FBN1, ZEB1, ZCCHC24, and LIFR. Overall, this demonstrates that these genes and pathways are the main driving forces behind tissue invasion and metastasis in UCS. \n",
    "\n",
    "This conclusion is supported by a variety of papers. FBN1, for example, is highlighted to lead to poor survival rates and high metastasis when there are high levels of expression in cases of ovarian cancer. Our conclusion extrapolates this paper to cases of UCS because even though research is limited, the patterns between more studied cancers and UCS should not be ignored. ZEB1 promotes metastasis because it is a transcription factor that leads to the promotion of EMT, which helps uterine cancer cells migrate to other parts of the body. ZEB1 can also contribute to the positive feedback loop that leads to more EMT production and thus more metastasis. There is little to no data that directly relates ZCCHC24 to UCS; however, it is related to metastasis in many other types of cancers due to its high expression in mesenchymal-like cancer stem cells. Similarly, LIFR is currently understudied with its relationship to UCS but is frequently present in aggressive types of endometrial cancers and is shown to help cancer cells survive and travel within the body/ immune system.\n",
    "\n",
    "Work cited:\n",
    "* NIH - The extracellular matrix glycoprotein fibrillin-1 in health and disease\n",
    "* NIH - ZEB1 Expression in Endometrial Biopsy Predicts Lymph Node Metastases in Patient with Endometrial Cancer\n",
    "* NIH - LIF/LIFR oncogenic signaling is a novel therapeutic target in endometrial cancer\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56d9e1e7",
   "metadata": {},
   "source": [
    "## Conclusions and Ethical Implications: \n",
    "*(Think about the answer your analysis generated, draw conclusions related to your overarching question, and disc uss the ethical implications of your conclusions.*  \n",
    "\n",
    "Conclusion: \n",
    "\n",
    "* The results of the analysis support the overarching goal of identifying the key genetic pathways involved in UCS development and metastasis. From the PCA, it revealed two major biological axes underlying tumor variability: an epithelial-loss axis (PC1) and a mesenchymal/invasion axis (PC2). The genres most strongly driving these axes are GRHL2, KRT8, CX3CL1, and CRB3 for epithelial structure, and FBN1, ZEB1, ZCCHC24, and LIFR for invasion and vascular remodeling. These highlight the transition from stable epithelial states to aggressive mesenchymal phenotypes that facilitate metastasis. These findings align with published evidence showing that genes like ZEB1 and FBN1 actively promote epithelial-mesenchymal transition (EMT), extracellular matrix remodeling, and dissemination in gynecologic cancers. Together this suggests that UCS tumors are not just becoming more invasive (through EMT pathways) but are also increasing their ability to grow and spread by forming new blood vessels (through angiogenesis pathways). However, because PCA is not a mechanistic test and the dataset is limited, these results cannot establish causation but instead point to possible pathways and gene targets that future research could explore for therapeutic development. \n",
    "\n",
    "\n",
    "Ethical Implications:  \n",
    "\n",
    "* For ethical limitation, two factors must be considered. First, PCA and clustering only show correlation, but cannot prove causation. Therefore, clinical recommendations cannot be made without further biological validation. Furthermore, since UCS is rare and understudied, applying insights from other cancers must be done cautiously in order to avoid overstating their relevance. Overall, the research has to be transparent about uncertainty, not overstating what the results can prove, and making sure that any potential drug targets are thoroughly tested in the lab and in clinical studies before they are considered for use in patients."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f023b735-6efb-43ed-a03d-eb4a9cdb734e",
   "metadata": {},
   "source": [
    "## Limitations and Future Work: \n",
    "*(Think about the answer your analysis generated, draw conclusions related to your overarching question, and discuss the ethical implications of your conclusions.* \n",
    "\n",
    "The limitation of this project is the small data set and the overall lack of research on UCS. We also don't have the background on how this data was collected, so the conclusions we can draw are very limited. Additionally, PCA is not a mechanistic test and therefore cannot make conclusions on HOW a gene influences something. With this being said, we cannot say that we know what causes metastasis, but there is a relationship between the expression of these genes and metastasis. \n",
    "\n",
    "Future work can provide a more explicit derivation of samples in order to better understand if these patterns of gene expression are in every tissue or are locally regulated. We can also look at genes outside of this gene set and across other types of cancer that have known metastasis-associated genes. This could include endometrial and uterine cancers. We could additionally look into unsupervised learning to see if there are ANY up-/down regulated genes in tumors of this kind, and then further look into what those genes are known for.  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9295960-2404-43dc-b46f-9a6f823f1657",
   "metadata": {},
   "source": [
    "## NOTES FROM YOUR TEAM: \n",
    "* the research on this cancer is limited, so some inference are being made fro gernal uterine cancer studies "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd725b7f-9741-46b6-a213-9518da9201c3",
   "metadata": {},
   "source": [
    "## QUESTIONS FOR YOUR TA: \n",
    "* what should we do if other genes that aren't in dataset have large effect on this cancer? "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
